perf(semantic): remove function stack

overlookmotel opened the pull request:
Only purpose of `function_stack` in `SemanticBuilder` is to track the current function, in order to set `NodeFlags::HasYield` on it if it contains `yield`.

#12630 moved enter and exit logic into `visit_function`. So now we don't need a `Stack` to track the current function, we can use vars on the actual stack.

Also, it's unnecessary to update current function when entering/exiting arrow functions, as they can't contain `yield`.

Dunqing reviewed (approved):
Looks great!

Dunqing commented:
### Merge activity

* **Aug 1, 9:45 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
perf(semantic): remove function stack (#12676)

Only purpose of `function_stack` in `SemanticBuilder` is to track the current function, in order to set `NodeFlags::HasYield` on it if it contains `yield`.

#12630 moved enter and exit logic into `visit_function`. So now we don't need a `Stack` to track the current function, we can use vars on the actual stack.

Also, it's unnecessary to update current function when entering/exiting arrow functions, as they can't contain `yield`.
